### PR TITLE
Add unit tests to blockchain reader

### PR DESCRIPTION
**Pull request by stephenctw**

### CODE REVIEW

File: cartesi-rollups/node/blockchain-reader/src/lib.rs
Comment (GCdePaula) [inline cartesi-rollups/node/blockchain-reader/src/lib.rs:1]: I see there are many ways in the codebase to launch anvil and deploy our smart contracts. Is there maybe a way to unify it?

I think the preferable way is through forge scripts. I believe it's possible to store the state of anvil and load it later:

```
anvil --dump-state state.json
anvil --load-state state.json
```

Just an idea, but maybe we could have a build target that generates this "template" (spawn anvil with `--dump-state`, runs forge script, exits), and all the tests that need an anvil+contracts just run `anvil --load-state state.json`.

File: cartesi-rollups/node/blockchain-reader/src/lib.rs [hunk 1]
```diff
@@ -438,32 +439,364 @@ impl PartitionProvider {
...
+    };
+
+    type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;
+    const APP_ADDRESS: Address = Address::ZERO;
+    const URL: &str = "http://127.0.0.1:8545";
+    const INITIAL_STATE: Digest = Digest::ZERO;
+
+    async fn spawn_anvil() -> Result<Child> {
+        // Start the Anvil node
+        let anvil = Command::new("anvil")
```
[the remark this answers is not in the document]
Comment (GCdePaula) [reply cartesi-rollups/node/blockchain-reader/src/lib.rs:486]: I think it could simplify a lot, such as removing `kill_child` etc.

**Review by GCdePaula — COMMENTED** [review]
Great work! I left a few comments.

File: cartesi-rollups/node/blockchain-reader/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (stephenctw) [reply cartesi-rollups/node/blockchain-reader/src/lib.rs:486]: Thanks for the suggestion, I'll look into that.

File: cartesi-rollups/node/blockchain-reader/src/lib.rs [hunk 2]
```diff
@@ -438,32 +439,364 @@ impl PartitionProvider {
...
+
+        // kill anvil should cause the blockchain_reader to crash, thus end the test
+        kill_child(anvil).unwrap();
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    // note: the tests involve interacting with anvil blockchain and a sqlite3 database, can be a bit time consuming (30+ seconds)
+    async fn test_readers_sequentially() -> Result<()> {
```
[the remark this answers is not in the document]
Comment (stephenctw) [reply cartesi-rollups/node/blockchain-reader/src/lib.rs:795]: I think I was trying to spawn a new anvil process for each test, but maybe that's not necessary, all tests can share one anvil node.

File: cartesi-rollups/node/blockchain-reader/src/lib.rs
Comment (stephenctw) [reply cartesi-rollups/node/blockchain-reader/src/lib.rs:1]: I really like this idea. I think this can be a separate task/PR from the current one.

**Review by GCdePaula — APPROVED** [review]
LGTM!

**Pull request merged**